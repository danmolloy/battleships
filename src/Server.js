import { createServer, Model } from "miragejs"

export function makeServer({ environment = "test" } = {}) {
  let server = createServer({
    environment,

    models: {
      user: Model,
    },

    seeds(server) {
      server.create("user", { name: "Bob", id: 1, moves: 48 })
      server.create("user", { name: "Alice", id: 2, moves: 34 })
    },

    routes() {
      this.namespace = "api"

      this.get("/users", (schema) => {
        return schema.users.all()
      })
    },
  })

  return server
}