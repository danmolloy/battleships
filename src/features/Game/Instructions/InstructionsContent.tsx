export const instructionsContent = [
  {text: `The aim of Battleships is to sink all of your opponents ships before they sink yours.`, 
  key: 1},
  {text: `There are two 10 * 10 boards which represent a space of ocean. \n
  One contains your ships, the other contains your opponent's.`, 
  key: 2},
  {text: `Parts of a ship are represented with a • \n
  There are small ships •• \n
  And large ships ••••• \n 
  And ships in between.`, key: 3},
  {text: `Upon clicking Start, the boards are both rendered with following ships: \n
  1 x Carrier (•••••) \n
  1 x Battleship (••••) \n
  1 x Destroyer (•••) \n
  1 x Submarine (•••) \n
  1 x Patrol Boat (••) \n`, key: 4},
  {text: `Players take turns attacking their opponents board, attempting to hit a ship. \n
  A hit is represented by X, a miss is O. \n
  Each player can only see their own ships.`, key: 5},
  {text:`If you hit, try the squares around that square to sink the entire ship. \n
  The ships may be horizontal or vertical. They cannot be touching another ship (including a diagonal square). \n
  You can see the ships remaining for both players, and will be notified when you sink one.`, key: 6}
]

