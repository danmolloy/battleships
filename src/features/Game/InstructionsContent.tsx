export const instructionsContent = [
  `The aim of Battleships is to sink all of your opponents ships before they sink yours.`,
  `There are two 10 * 10 boards which represent a space of ocean. \n
  One contains your ships, the other contains your opponent's.`,
  `Parts of a ship are represented with a • \n
  There are small ships •• \n
  And large ships ••••• \n 
  And ships in between.`,
  `Upon clicking Start, the boards are both rendered with following ships: \n
  1 x Carrier (•••••) \n
  1 x Battleship (••••) \n
  1 x Destroyer (•••) \n
  1 x Submarine (•••) \n
  1 x Patrol Boat (••) \n
  Each player can only see their own ships.`,
  `Players take turns attacking their opponents board, attempting to hit a ship. \n
  A hit is represented by X, a miss is O.`,
  `If you hit, try the squares around that square to sink the entire ship. \n
  The ships may be horizontal or vertical. They cannot be touching another ship (including a diagonal square). \n
  You can see the ships remaining for both players, and will be notified when you sink one.`
]

