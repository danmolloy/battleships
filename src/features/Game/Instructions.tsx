import { useState } from "react"
import { instructionsContent } from "./InstructionsContent"

export const Instructions = (props: any) => {
  const [pageNumber, setPageNumber] = useState(1)

  const handleClick = () => {
    pageNumber < instructionsContent.length
    ? setPageNumber(pageNumber + 1)
    : props.close()
  }

  return (
    <div className="flex flex-col border md:w-1/2 bg-white fixed text-left h-auto top-2 sm:top-12 sm:left-20 sm:right-20 left-4 right-4 shadow">
      <div className="flex flex-row items-end self-end w-full p-2">
        <h2 className="self-center w-3/4">Instructions</h2>
        <p className="w-1/4">{`${pageNumber}/${instructionsContent.length}`}</p>
      </div>
      <div className="my-4">
      {instructionsContent[pageNumber - 1].split('\n').map((i: any) => 
      <p className="p-2">{i}</p>)}
      </div>
      <div className="flex flex-row justify-end w-full">
        <button onClick={() => handleClick()} className="self-end text-blue-500 hover:text-blue-700">
          <svg xmlns="http://www.w3.org/2000/svg" className="h-12 w-12" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M13 7l5 5m0 0l-5 5m5-5H6" />
          </svg>
        </button>
      </div>
    </div>
  )
}