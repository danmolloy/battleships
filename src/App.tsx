import * as React from 'react';
import './App.css';
import { CPUShips } from './features/cpuBoard/CPUShips';
import { Header } from './features/Game/Header'
import { PlayerShips } from './features/playerBoard/PlayerShips';

function App() {
  return (
    <div 
    className="flex flex-col justify-center items-center text-center"
    >
      <Header />
      <div className="flex flex-col sm:flex-row">
      <CPUShips />
      <PlayerShips />
      </div>
    </div>
  );
}

export default App;
