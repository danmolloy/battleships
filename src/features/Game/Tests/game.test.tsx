import { fireEvent, getByText } from '@testing-library/react';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils"
import App from '../../../App';
import { Provider } from 'react-redux';
import pretty from 'pretty';
import { configureStore } from '@reduxjs/toolkit';
import cpuShipsReducer, { resetCPUBoard } from '../../cpuBoard/cpuShipsSlice'
import playerShipsReducer from '../../playerBoard/playerShipsSlice'
import gameReducer, { addPlayerData } from '../../Game/gameSlice'
import { instructionsContent } from '../Instructions/InstructionsContent'
import { makeServer } from '../../../Server'

jest.spyOn(window, 'alert').mockImplementation(() => {});

let container: any = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove();
  container = null;
});

describe("App", () => {

  it('renders app without error', () => {
    act(() => {
      render(
      <Provider store={configureStore({
        reducer: {
          game: gameReducer,
          cpuShips: cpuShipsReducer,
          playerShips: playerShipsReducer
        }
      })}>
        <App />
      </Provider>, container)
    })
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  })

  it('handles Start button click', () => {
    act(() => {
      render( 
        <Provider store={configureStore({
          reducer: {
            game: gameReducer,
            cpuShips: cpuShipsReducer,
            playerShips: playerShipsReducer
          }
        })}>
          <App />
        </Provider>, container
      )
    })
    fireEvent.click(getByText(container, "Start"));

    expect(container.querySelector("#cpu-board-inner").textContent.length).toBe(0);
    expect(container.querySelector("#cpu-board-inner").textContent).not.toMatch(/•/g);

    expect(pretty(container.querySelector("#CPU-board").innerHTML)).toMatchSnapshot();

    expect(container.querySelector("#player-board-inner").textContent.length).toBe(17);
    expect(container.querySelector("#player-board-inner").textContent).toMatch(/^•{17}$/);
    expect(pretty(container.querySelector("#Player-board").innerHTML)).toMatchSnapshot();

    expect(container.textContent).toMatch(/Player turn/);
    expect(container.querySelector("#status-btn").textContent).toMatch(/^Reset$/)
  })

  it("handles player turn and makes CPU turn", async () => {
    act(() => {
      render( 
        <Provider store={configureStore({
          reducer: {
            game: gameReducer,
            cpuShips: cpuShipsReducer,
            playerShips: playerShipsReducer
          }
        })}>
          <App />
        </Provider>, container
      )
    })

    fireEvent.click(getByText(container, "Start"));
    let randInt = Math.floor(Math.random() * 100)

    await act(async () => {
      fireEvent.click(container.querySelector("#cpu-board-inner").children[randInt])
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    
    expect(container.querySelector("#cpu-board-inner").children[randInt].textContent).toMatch(/^[OX]$/)
    expect(container.querySelector("#Player-board").textContent).toMatch(/Attack count: 1/g)
    expect(container.textContent).toMatch(/CPU turn/g)
    
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 1000))
    })
    expect(container.querySelector("#player-board-inner").textContent).toMatch(/^•*[OX]•*$/)
    expect(container.textContent).toMatch(/Player turn/g)
    expect(container.querySelector("#CPU-board").textContent).toMatch(/Attack count: 1/g)
  })


  it("calls reload on Reset button click", () => {
    Object.defineProperty(window, 'location', {
      value: { reload: jest.fn() }
    });
    act(() => {
      render( 
        <Provider store={configureStore({
          reducer: {
            game: gameReducer,
            cpuShips: cpuShipsReducer,
            playerShips: playerShipsReducer
          }
        })}>
          <App />
        </Provider>, container
      )
    })
    act(() => {
      fireEvent.click(getByText(container, "Start"));
    })
    expect(container.textContent).toMatch(/Reset/g)
    act(() => {
      fireEvent.click(getByText(container, "Reset"));
    })
    expect(window.location.reload).toHaveBeenCalled();
    
  })

  it("renders instructions", () => {
    act(() => {
      render( 
        <Provider store={configureStore({
          reducer: {
            game: gameReducer,
            cpuShips: cpuShipsReducer,
            playerShips: playerShipsReducer
          }
        })}>
          <App />
        </Provider>, container
      )
    })
    act(() => {
      fireEvent.click(getByText(container, "Instructions"));
    })
    
    expect(container.textContent).toMatch(instructionsContent[0].text)
    expect(container.textContent).toMatch(/1\/6/g)

    act(() => {
      fireEvent.click(container.querySelector("#next-instruction-btn"))
    })
    expect(container.textContent).toMatch(instructionsContent[1].text.slice(0,20))
    expect(container.textContent).toMatch(/2\/6/g)

    act(() => {
      fireEvent.click(container.querySelector("#next-instruction-btn"))
    })
    expect(container.textContent).toMatch(instructionsContent[2].text.slice(0,20))
    expect(container.textContent).toMatch(/3\/6/g)

    act(() => {
      fireEvent.click(container.querySelector("#next-instruction-btn"))
    })
    expect(container.textContent).toMatch(instructionsContent[3].text.slice(0,20))
    expect(container.textContent).toMatch(/4\/6/g)

    act(() => {
      fireEvent.click(container.querySelector("#next-instruction-btn"))
    })
    expect(container.textContent).toMatch(instructionsContent[4].text.slice(0,20))
    expect(container.textContent).toMatch(/5\/6/g)

    act(() => {
      fireEvent.click(container.querySelector("#next-instruction-btn"))
    })
    expect(container.textContent).toMatch(instructionsContent[5].text.slice(0,20))
    expect(container.textContent).toMatch(/6\/6/g)
    act(() => {
      fireEvent.click(container.querySelector("#next-instruction-btn"))
    })
    expect(container.textContent).not.toMatch(/\/6/g)
  })

  it("fetches Highscores", async () => {
    makeServer({ environment: "test" })

    act(() => {
      render( 
        <Provider store={configureStore({
          reducer: {
            game: gameReducer,
            cpuShips: cpuShipsReducer,
            playerShips: playerShipsReducer
          }
        })}>
          <App />
        </Provider>, container
      )
    })

    await act(async () => {
      fireEvent.click(container.querySelector("#highscores-btn"))
      await new Promise(resolve => setTimeout(resolve, 1100))
    })
    expect(container.textContent).toMatch(/Kelly38Eoghan44Fiona45Ed48Dan49Jorge54/g);
    expect(container.textContent).toMatchSnapshot();
    
  })

  it("handles end of game without error", async() => {
    let store = configureStore({
      reducer: {
        game: gameReducer,
        cpuShips: cpuShipsReducer,
        playerShips: playerShipsReducer
      }
    })
    act(() => {
      render( 
        <Provider store={store}>
          <App />
        </Provider>, container
      )
    })

    act(() => {
      fireEvent.click(getByText(container, "Start"))
    })

    expect(store.getState().game.inGame).toBe("playing")

    await act(async() => {
      store.dispatch(resetCPUBoard())
      await new Promise(resolve => setTimeout(resolve, 1200))
    })

    expect(container.textContent).toMatch(/Kelly38Eoghan44Fiona45Ed48Dan49Jorge54/g);
    expect(container.textContent).toMatch(/Your name:Submit/g);
    expect(container.textContent).toMatch(/Ships Remaining: 0/g);

  })

  it("submits highscore", async () => {
    
    let store = configureStore({
      reducer: {
        game: gameReducer,
        cpuShips: cpuShipsReducer,
        playerShips: playerShipsReducer
      }
    })
    act(() => {
      render( 
        <Provider store={store}>
          <App />
        </Provider>, container
      )
    })

    act(() => {
      fireEvent.click(getByText(container, "Start"))
    })

    expect(store.getState().game.inGame).toBe("playing")

    await act(async() => {
      store.dispatch(resetCPUBoard())
      await new Promise(resolve => setTimeout(resolve, 1200))
    })

    expect(container.textContent).toMatch(/Kelly38Eoghan44Fiona45Ed48Dan49Jorge54Your name:Submit/g)

    await act(async () => {
      store.dispatch(addPlayerData({name: "Jest", moves: 100}))
      await new Promise(resolve => setTimeout(resolve, 1200))
    })
    expect(container.textContent).toMatch(/Jorge54Jest100/g)
  })

})