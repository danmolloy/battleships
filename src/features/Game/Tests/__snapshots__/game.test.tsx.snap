// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App fetches Highscores 1`] = `"BattleshipsStartInstructionsHigh ScoresNameMovesKelly38Eoghan44Fiona45Ed48Dan49Jorge54CPU ShipsPlayer Ships"`;

exports[`App handles Start button click 1`] = `
"<p>Attack count: 0</p>
<div>
  <div>
    <div>
      <p>Ships Remaining: 5</p><i class=\\"up-arrow\\"></i>
    </div>
    <p>(2) PatrolBoat</p>
    <p>(3) Destroyer</p>
    <p>(3) Submarine</p>
    <p>(4) Battleship</p>
    <p>(5) Carrier</p>
  </div>
</div>"
`;

exports[`App handles Start button click 2`] = `
"<p>Attack count: 0</p>
<div>
  <div>
    <div>
      <p>Ships Remaining: 5</p><i class=\\"up-arrow\\"></i>
    </div>
    <p>(2) PatrolBoat</p>
    <p>(3) Destroyer</p>
    <p>(3) Submarine</p>
    <p>(4) Battleship</p>
    <p>(5) Carrier</p>
  </div>
</div>"
`;

exports[`App renders app without error 1`] = `
"<div class=\\"flex flex-col justify-center items-center text-center\\">
  <div class=\\"header m-3\\" id=\\"header\\">
    <h1 id=\\"title\\" class=\\"p-4\\">Battleships</h1>
    <div class=\\"flex flex-col sm:flex-row items-center\\"><button id=\\"status-btn\\" class=\\"header-btn bg-green-500 text-white hover:bg-green-600 focus:ring-green-400\\">Start</button><button class=\\"header-btn bg-blue-500 text-white hover:bg-blue-600 focus:ring-gray-500\\">Instructions</button><button id=\\"highscores-btn\\" class=\\"header-btn bg-gray-200 text-gray-600 hover:bg-gray-300 focus:ring-gray-500\\">High Scores</button></div>
  </div>
  <div class=\\"flex flex-col sm:flex-row items-start\\">
    <div class=\\"outer-board pb-1\\" id=\\"cpu-ships\\">
      <div id=\\"cpu-board-inner\\" class=\\"inner-board shadow-sm\\">
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
        <div class=\\"square hover:bg-blue-300\\"></div>
      </div>
      <h3>CPU Ships</h3>
    </div>
    <div class=\\"outer-board pb-1\\" id=\\"player-ships\\">
      <div id=\\"player-board-inner\\" class=\\"inner-board\\">
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
        <div class=\\"square\\"></div>
      </div>
      <h3>Player Ships</h3>
    </div>
  </div>
</div>"
`;
